Add Contact.get_merge_conflicts api

eileenmcnaughton opened the pull request:
Overview
----------------------------------------
Cleans up code & exposed Contact.get_merge_conflicts as a merge action

Before
----------------------------------------
Less testing, less readability, no ability to access via api

After
----------------------------------------
API:
```
civicrm_api3('Contact', 'get_merge_conflicts', ['to_keep_id => 1, 'to_remove_id] => 2);
```
Technical Details
----------------------------------------
The goal here is to
1) improve testing
2) cleanup the code & define parameters where they are actually used rather than pass from pillar to post
3) expose to the api so we can be less coupled to the current form

I thought about putting the api on the Dedupe entity but my thinking is that actions relating to
individual pairs go on the contact entity & those relating to the batch find, cache, merge
functionality go on the Dedupe entity.

I'm doing another round of dedupe code cleanup at the moment... In conjunction with efforts to write an
angular form which highlights what functionality a form needs to access & where the code is awful....


Comments
----------------------------------------
@JKingsnorth

pfigel commented on api/v3/Contact.php:
```diff
@@ -1213,22 +1213,87 @@ function civicrm_api3_contact_merge($params) {
+  ];
+  $params['mode'] = [
+    'title' => ts('Dedupe mode'),
+    'description' => ts("'safe' or 'aggressive'  - these modes map to the merge actions & may affect resolution done by hooks "),
```
This could use an `option` with the allowed values.

eileenmcnaughton commented:
@pfigel I've thought a bit about your comment & I concluded that we actually want a multi-tiered array.  

I've changed it so the results look more like
```
      'safe' => [
        'conflicts' => [
          'contact' => [['first_name' => [3 => 'Anthony', 4 => 'different']]],
          'address' => [
            [
              'location_type_id' => '1',
              'street_address' => [
                3 => 'big house',
                4 => 'medium house',
              ],
            ],
            [
              'location_type_id' => '2',
              'street_address' => [
                3 => 'big office',
                4 => 'medium office',
              ],
            ],
          ],
          'email' => [
            [
              'location_type_id' => '1',
              'email' => [
                3 => 'bob@example.com',
                4 => 'anthony_anderson@civicrm.org',
              ],
            ],
          ],
        ],
      ],
```
I felt like any munging the locations into flat keys was ultimately going to lead to the sort of mess we have in the BAO layer already. Since I added multiple keys for the various entities I thought it made sense to have contact as an array too for more consistency.

I also added some outer layers - ie. I keyed by mode & added a 'conflicts' key. The reason I did that is I can imagine wanting to add more keys - that might be metadata or it might be 'resolutions' - ie. what will this conflict resolve as if you force it. Note that you can affect that by hook depending on mode so I'm imagining for us 'safe mode' would be a conflict on first name whereas in aggressive mode we would ideally resolve to most recent donor.

This thinking has come out of trying to imagine a different, non quickform!, form to speed up the process of manual deduping. On our site conflicts without conflicts would already be merged but for those with conflicts a quick eyeball is enough to click on 'force merge' -  ie in the screen shot below they can choose 'force merge' to accept resolving the conflict by keeping the first name 'bobppp'

<img width="1137" alt="Screen Shot 2019-06-03 at 3 47 50 PM" src="https://user-images.githubusercontent.com/336308/58774723-fd626f80-8616-11e9-8008-42a7cdf6b689.png"


All of which is a bit of a long background  but I wanted to explain how I think this api would be used - other than to clean up the code. My big picture goal here is to expose the workings of our dedupe code to the api so we can swap out the form layer

eileenmcnaughton commented:
test this please

eileenmcnaughton commented:
test this please

eileenmcnaughton commented:
test this please

eileenmcnaughton commented:
@pfigel do you have any more thoughts on this? It would be nice if this (& maybe https://github.com/civicrm/civicrm-core/pull/14411 ) could hit the rc

pfigel commented:
@eileenmcnaughton the new array structure definitely seems better/easier to work with. I did some further testing and there's some weird behaviour where, given a certain number of conflicts, the values start disappearing. I started with just a conflict on `last_name`, that worked:

```json
"safe": {
    "conflicts": {
        "contact": [
            {
                "last_name": {
                    "204": "Doe",
                    "207": "Doee"
                }
            }
        ]
    }
}
```

As soon as I add a conflict on e.g. `external_identifier`, that value disappears:

```json
"safe": {
    "conflicts": {
        "contact": [
            {
                "external_identifier": [],
                "last_name": {
                    "204": "Doee",
                    "207": "Doee"
                }
            }
        ]
    }
}
```

With a conflict *only* on `external_identifier`, however, it works:

```json
"safe": {
    "conflicts": {
        "contact": [
            {
                "external_identifier": {
                    "204": "4444X",
                    "207": "123T"
                }
            }
        ]
    }
}
```

For custom fields, it seems to be missing the values for the `to_remove_id` contact (the contacts have a non-empty, differing value for this custom field):

```json
"safe": {
   "conflicts": {
       "contact": [
           {
               "custom_2": {
                   "204": "S",
                   "207": null
               }
           }
       ]
   }
}
```

Not sure if all of that is related to this PR or just general brokenness in the existing code the API utilizes.

eileenmcnaughton commented:
@pfigel the things you pointed out were in the new code. I extended the tests to capture them & then fixed

pfigel commented:
@eileenmcnaughton looks better now - no empty arrays anymore. There's still some weirdness with values disappearing for the `to_remove_id` contact if there's a conflict on a custom field:

```json
"safe": {
    "conflicts": {
        "contact": [
            {
                "external_identifier": {
                    "204": "4444X",
                    "207": null
                },
                "last_name": {
                    "204": "Figell",
                    "207": null
                },
                "custom_2": {
                    "204": "S",
                    "207": null
                }
            }
        ]
    }
}
```
If I resolve the conflict on `custom_2`, it works correctly.

eileenmcnaughton commented:
@pfigel do you have civicase enabled? I don't normally work much with CiviCase but I have it enabled now for some testing & I find that if I add a custom field to the contact.get call it goes & adds an inner jojn!

```
SELECT contact_a.id as contact_id, contact_a.external_identifier as `external_identifier`, contact_a.first_name as `first_name`, contact_a.last_name as `last_name`, civicrm_value_volunteer_com_6.id as civicrm_value_volunteer_com_6_id, civicrm_value_volunteer_com_6.volunteer_project_id_12 as custom_12, civicrm_value_volunteer_inf_5.id as civicrm_value_volunteer_inf_5_id, civicrm_value_volunteer_inf_5.camera_skill_level_11 as custom_11, civicrm_value_volunteer_inf_5.country_13 as custom_13, civicrm_value_volunteer_inf_5.file_14 as custom_14,
       civicrm_value_constituent_information_1.id as civicrm_value_constituent_information_1_id, civicrm_value_constituent_information_1.most_important_issue_1 as custom_1, civicrm_value_constituent_information_1.marital_status_2 as custom_2, civicrm_value_constituent_information_1.marriage_date_3 as custom_3, civicrm_value_donor_information_3.id as civicrm_value_donor_information_3_id, civicrm_value_donor_information_3.how_long_have_you_been_a_donor_6 as custom_6, civicrm_value_donor_information_3.known_areas_of_interest_5 as custom_5, civicrm_value_food_preference_2.id as civicrm_value_food_preference_2_id, civicrm_value_food_preference_2.soup_selection_4 as custom_4, civicrm_value_civivolunteer_4.id as civicrm_value_civivolunteer_4_id, civicrm_value_civivolunteer_4.volunteer_need_id_7 as custom_7, civicrm_value_civivolunteer_4.time_scheduled_in_minutes_9 as custom_9, civicrm_value_civivolunteer_4.volunteer_role_id_8 as custom_8, civicrm_value_civivolunteer_4.time_completed_in_minutes_10 as custom_10, civicrm_value_status_7.id as civicrm_value_status_7_id, civicrm_value_status_7.required_15 as custom_15, civicrm_value_status_7.found_16 as custom_16,
       civicrm_value_lego_piece_de_8.id as civicrm_value_lego_piece_de_8_id, civicrm_value_lego_piece_de_8.field_id_17 as custom_17, civicrm_value_lego_piece_de_8.colour_18 as custom_18, civicrm_value_lego_piece_de_8.length_19 as custom_19, civicrm_value_lego_piece_de_8.width_20 as custom_20, civicrm_value_lego_piece_de_8.description_21 as custom_21, civicrm_value_lego_piece_de_8.is_technic__22 as custom_22, civicrm_value_memberships_9.id as civicrm_value_memberships_9_id, civicrm_value_memberships_9.memb_23 as custom_23, civicrm_value_activity_fiel_10.id as civicrm_value_activity_fiel_10_id, civicrm_value_activity_fiel_10.acgivity_fiels_24 as custom_24, civicrm_value_prospect_financial_information.id as civicrm_value_prospect_financial_information_id,
       civicrm_value_prospect_financial_information.campaign_id as custom_25, civicrm_value_prospect_financial_information.budgeted as custom_26, civicrm_value_prospect_financial_information.prospect_amount as custom_27, civicrm_value_prospect_financial_information.probability as custom_28, civicrm_value_prospect_financial_information.expectation as custom_29, civicrm_value_prospect_financial_information.expected_date as custom_30, civicrm_value_prospect_financial_information.restricted_code as custom_33
FROM civicrm_contact contact_a 
LEFT JOIN civicrm_contribution ON civicrm_contribution.contact_id = contact_a.id
LEFT JOIN civicrm_participant ON civicrm_participant.contact_id = contact_a.id
 LEFT JOIN civicrm_case_contact ON civicrm_case_contact.contact_id = contact_a.id
  
 INNER JOIN civicrm_case ON civicrm_case_contact.case_id = civicrm_case.id
    LEFT JOIN civicrm_activity_contact

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         ON ( civicrm_activity_contact.contact_id = contact_a.id )
    LEFT JOIN civicrm_activity
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     ON ( civicrm_activity.id = civicrm_activity_contact.activity_id
AND civicrm_activity.is_deleted = 0 AND civicrm_activity.is_current_revision = 1 ) 
INNER JOIN civicrm_contact
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                       ON ( civicrm_activity_contact.contact_id = civicrm_contact.id and civicrm_contact.is_deleted != 1 ) 

LEFT JOIN civicrm_membership ON civicrm_membership.contact_id = contact_a.id  LEFT JOIN civicrm_contribution_recur ccr ON ( civicrm_membership.contribution_recur_id = ccr.id ) 
LEFT JOIN civicrm_relationship ON (civicrm_relationship.contact_id_a = contact_a.id ) 
LEFT JOIN civicrm_contact contact_b ON (civicrm_relationship.contact_id_b = contact_b.id )
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                             LEFT JOIN civicrm_value_activity_fiel_10 ON civicrm_value_activity_fiel_10.entity_id = `civicrm_activity`.id

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                             LEFT JOIN civicrm_value_civivolunteer_4 ON civicrm_value_civivolunteer_4.entity_id = `civicrm_activity`.id
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                             LEFT JOIN civicrm_value_constituent_information_1 ON civicrm_value_constituent_information_1.entity_id = `contact_a`.id
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           LEFT JOIN civicrm_value_donor_information_3 ON civicrm_value_donor_information_3.entity_id = `civicrm_contribution`.id
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                             LEFT JOIN civicrm_value_food_preference_2 ON civicrm_value_food_preference_2.entity_id = `civicrm_participant`.id

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                             LEFT JOIN civicrm_value_lego_piece_de_8 ON civicrm_value_lego_piece_de_8.entity_id = `contact_a`.id
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                             LEFT JOIN civicrm_value_memberships_9 ON civicrm_value_memberships_9.entity_id = `civicrm_membership`.id
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                             LEFT JOIN civicrm_value_prospect_financial_information ON civicrm_value_prospect_financial_information.entity_id = `civicrm_case`.id
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                             LEFT JOIN civicrm_value_status_7 ON civicrm_value_status_7.entity_id = `civicrm_relationship`.id
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                             LEFT JOIN civicrm_value_volunteer_com_6 ON civicrm_value_volunteer_com_6.entity_id = `civicrm_activity`.id
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                             LEFT JOIN civicrm_value_volunteer_inf_5 ON civicrm_value_volunteer_inf_5.entity_id = `contact_a`.id  WHERE  ( contact_a.id IN ("217", "216") )


```

eileenmcnaughton commented:
test this please

pfigel commented:
@eileenmcnaughton close enough, it wasn't CiviCase, but there's something that adds this to the query as soon as I add a custom field:
```sql
LEFT JOIN civicrm_activity ON ( civicrm_activity.id = civicrm_activity_contact.activity_id AND civicrm_activity.is_deleted = 0 AND civicrm_activity.is_current_revision = 1 )
INNER JOIN civicrm_contact ON ( civicrm_activity_contact.contact_id = civicrm_contact.id and civicrm_contact.is_deleted != 1 )
```

One of the test contacts didn't have *any* activities, so that would get it to drop.

I was able to reproduce this against dmaster, but not 5.14 or 5.13, so it looks like an unreleased regression not related to this PR. I'll file a ticket.

eileenmcnaughton commented:
@pfigel ok - I'll merge this then since it's something different
